lib/src/yaml_edit/editor.dart

```diff
@@ -0,0 +1,679 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection' show UnmodifiableListView;
+
+import 'package:meta/meta.dart';
+import 'package:yaml/yaml.dart';
+
+import 'equality.dart';
+import 'errors.dart';
+import 'list_mutations.dart';
+import 'map_mutations.dart';
+import 'source_edit.dart';
+import 'strings.dart';
+import 'utils.dart';
+import 'wrap.dart';
+
+/// An interface for modififying [YAML][1] documents while preserving comments
+/// and whitespaces.
+///
+/// YAML parsing is supported by `package:yaml`, and modifications are performed
+/// as string operations. Each time a modification takes place via one of the
+/// public methods, we calculate the expected final result, and parse the result
+/// YAML string, and ensure the two YAML trees match, throwing an
+/// [AssertionError] otherwise. Such a situation should be extremely rare, and
+/// should only occur with degenerate formatting.
+///
+/// Most modification methods require the user to pass in an [Iterable<Object>]
+/// path that holds the keys/indices to navigate to the element.
+///
+/// For example,
+///
+/// ```yaml
+/// a: 1
+/// b: 2
+/// c:
+///   - 3
+///   - 4
+///   - {e: 5, f: [6, 7]}
+/// ```
+///
+/// To get to `7`, our path will be `['c', 2, 'f', 1]`. The path for the base
+/// object is the empty array `[]`. All modification methods will throw an
+/// [Error] if the path provided is invalid. Note also that that the order of
+/// elements in the path is important, and it should be arranged in order of
+/// calling, with the first element being the first key or index to be called.
+///
+/// To dump the YAML after all the modifications have been completed, simply
+/// call [toString()].
+///
+/// [1]: https://yaml.org/
+@sealed
+class YamlEditor {
+  /// List of [SourceEdit]s that have been applied to [_yaml] since the creation
+  /// of this instance, in chronological order. Intended to be compatible with
+  /// `package:analysis_server`.
+  final List<SourceEdit> _edits = [];
+
+  UnmodifiableListView<SourceEdit> get edits =>
+      UnmodifiableListView<SourceEdit>([..._edits]);
```

Comment:
hmm... I shifted the documentation for `_edits` over to this function.

With regards to examples for how a list of source edits can be applied to a string, I think it might be better if I added the examples in the `source_edit.dart` instead!

Reply:
That's a good idea...

Reply:
hmm, that said... it might be nice to show the purpose of this property.

Or at-least say something like: The [SourceEdit] objects can be serialized to JSON using ...., deserialized using.... and applied to a string using `[SourceEdit.apply].